compiler/rustc_interface/src/queries.rs

```diff
@@ -19,43 +19,71 @@ use rustc_session::{output::find_crate_name, Session};
 use rustc_span::symbol::sym;
 use rustc_span::Symbol;
 use std::any::Any;
-use std::cell::{Ref, RefCell, RefMut};
+use std::cell::{Ref, RefCell};
 use std::rc::Rc;
 use std::sync::Arc;
 
 /// Represent the result of a query.
 ///
-/// This result can be stolen with the [`take`] method and generated with the [`compute`] method.
+/// This result can be stolen once with the [`take`] method and generated with the [`compute`] method.
 ///
 /// [`take`]: Self::take
 /// [`compute`]: Self::compute
 pub struct Query<T> {
-    result: RefCell<Option<Result<T>>>,
+    // The outer option is `None` until the query is computed.
+    // Then the whole thing becomes `Some(Some(result))`.
+    // Once the value gets taken/stolen, it is `Some(None)`.
+    result: RefCell<Option<Option<Result<T>>>>,
```

Comment:
At this point it's better to introduce a separate `enum { Uninitialized, Initialized(Value), Stolen }`.

Reply:
Yea I tried that but the code didn't really get simpler. I can push a version of it for comparison